pubsub/azure/eventhubs/eventhubs.go

```diff
@@ -98,62 +117,304 @@ func subscribeHandler(ctx context.Context, topic string, e *eventhub.Event, hand
+	StorageContainerName   string `json:"storageContainerName,omitempty"`
+	EnableEnitiyManagement bool   `json:"enableEntityManagement,omitempty,string"`
+	MessageRetentionInDays int32  `json:"messageRetentionInDays,omitempty,string"`
+	PartitionCount         int32  `json:"partitionCount,omitempty,string"`
+	SubscriptionID         string `json:"subscriptionID,omitempty"`
+	ResourceGroupName      string `json:"resourceGroupName,omitempty"`
 }
 
 // NewAzureEventHubs returns a new Azure Event hubs instance.
 func NewAzureEventHubs(logger logger.Logger) *AzureEventHubs {
 	return &AzureEventHubs{logger: logger}
 }
 
-func parseEventHubsMetadata(meta pubsub.Metadata) (azureEventHubsMetadata, error) {
+func parseEventHubsMetadata(meta pubsub.Metadata) (*azureEventHubsMetadata, error) {
+	b, err := json.Marshal(meta.Properties)
+	if err != nil {
+		return nil, err
+	}
+
 	m := azureEventHubsMetadata{}
+	err = json.Unmarshal(b, &m)
+	if err != nil {
+		return nil, err
+	}
 
-	if val, ok := meta.Properties[connectionString]; ok && val != "" {
-		m.connectionString = val
-	} else {
-		return m, errors.New(missingConnectionStringErrorMsg)
+	if m.ConnectionString == "" && m.EventHubNamespace == "" {
+		return &m, errors.New(missingConnectionStringNamespaceErrorMsg)
 	}
 
-	if val, ok := meta.Properties[storageAccountName]; ok && val != "" {
-		m.storageAccountName = val
-	} else {
-		return m, errors.New(missingStorageAccountNameErrorMsg)
+	if m.ConnectionString != "" && m.EventHubNamespace != "" {
+		return &m, errors.New(bothConnectionStringNamespaceErrorMsg)
 	}
 
-	if val, ok := meta.Properties[storageAccountKey]; ok && val != "" {
-		m.storageAccountKey = val
-	} else {
-		return m, errors.New(missingStorageAccountKeyErrorMsg)
+	return &m, nil
+}
+
+func validateAndGetHubName(connectionString string) (string, error) {
+	parsed, err := conn.ParsedConnectionFromStr(connectionString)
+	if err != nil {
+		return "", err
+	}
+	if parsed.HubName != "" {
+		return parsed.HubName, nil
 	}
+	return "", nil
+}
 
-	if val, ok := meta.Properties[storageContainerName]; ok && val != "" {
-		m.storageContainerName = val
-	} else {
-		return m, errors.New(missingStorageContainerNameErrorMsg)
+func (aeh *AzureEventHubs) ensureEventHub(hubName string) error {
+	if aeh.hubManager == nil {
+		aeh.logger.Errorf("hubManager client not initialized properly.")
+		return fmt.Errorf("hubManager client not initialized properly")
+	}
+	entity, err := aeh.getHubEntity(hubName)
+	if err != nil {
+		return err
+	}
+	if entity == nil {
+		if err := aeh.createHubEntity(hubName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (aeh *AzureEventHubs) ensureSubscription(hubName string) error {
+	err := aeh.ensureEventHub(hubName)
+	if err != nil {
+		return err
 	}
+	_, err = aeh.getConsumerGroupsClient()
+	if err != nil {
+		return err
+	}
+	return aeh.createConsumerGroup(hubName)
+}
+
+func (aeh *AzureEventHubs) getConsumerGroupsClient() (*mgmt.ConsumerGroupsClient, error) {
```

Comment:
consumer group client from the event hubs management clients in azure-go-sdk for managing consumer groups.